searchcore/src/vespa/searchcore/proton/server/lid_space_compaction_job_take2.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "lid_space_compaction_job_take2.h"
+#include "i_document_scan_iterator.h"
+#include "i_lid_space_compaction_handler.h"
+#include "i_operation_storer.h"
+#include <vespa/searchcore/proton/feedoperation/moveoperation.h>
+#include <vespa/persistence/spi/bucket_tasks.h>
+#include <vespa/vespalib/util/destructor_callbacks.h>
+#include <cassert>
+
+using search::DocumentMetaData;
+using search::LidUsageStats;
+using storage::spi::makeBucketTask;
+using storage::spi::Bucket;
+
+namespace proton::lidspace {
+
+bool
+CompactionJob::scanDocuments(const LidUsageStats &stats)
+{
+    if (_scanItr->valid()) {
+        DocumentMetaData document = getNextDocument(stats, false);
+        if (document.valid()) {
+            Bucket metaBucket(document::Bucket(_bucketSpace, document.bucketId));
+            IDestructorCallback::SP context = _moveOpsLimiter->beginOperation();
+            auto failed = _bucketExecutor.execute(metaBucket, makeBucketTask([this, meta=document, opsTracker=std::move(context)] (const Bucket & bucket, std::shared_ptr<IDestructorCallback> onDone) {
+                assert(bucket.getBucketId() == meta.bucketId);
+                using DoneContext = vespalib::KeepAlive<std::pair<IDestructorCallback::SP, IDestructorCallback::SP>>;
+                moveDocument(meta, std::make_shared<DoneContext>(std::make_pair(std::move(opsTracker), std::move(onDone))));
+            }));
+            if (failed) return false;
+            if (isBlocked(BlockedReason::OUTSTANDING_OPS)) {
+                return true;
+            }
+        }
+    }
+    return scanDocumentsPost();
+}
+
+void
+CompactionJob::moveDocument(const search::DocumentMetaData & metaThen, std::shared_ptr<IDestructorCallback> context) {
```

Comment:
This will be executed in the bucket executor thread. Is that safe? Compared to how this is currently handled in the LidSpaceCompactionJob.

Reply:
No, it is not, I will wire in forwarding to master executor when I wire it in. I have not agreed with myself on how yet.